sdks/python/apache_beam/runners/interactive/display/pipeline_graph.py

```diff
@@ -136,14 +159,50 @@ def _generate_graph_dicts(self):
           vertex_dict[invisible_leaf] = {'style': 'invis'}
           self._edge_to_vertex_pairs[pcoll_id].append(
               (transform.unique_name, invisible_leaf))
-          edge_dict[(transform.unique_name, invisible_leaf)] = {}
+          if self._pin:
+            edge_label = {'label':
+                            self._pin.cacheable_var_by_pcoll_id(pcoll_id)}
+            edge_dict[(transform.unique_name, invisible_leaf)] = edge_label
+          else:
+            edge_dict[(transform.unique_name, invisible_leaf)] = {}
+        # For PCollections with more than one consuming PTransform, we also add
+        # an invisible dummy node to diverge the edge in the middle as the
+        # single output is used by multiple down stream PTransforms as inputs
+        # instead of emitting multiple edges.
+        elif len(self._consumers[pcoll_id]) > 1:
+          intermediate_dummy = 'diverge{}'.format(
+              hash(pcoll_id) % 10000)
+          vertex_dict[intermediate_dummy] = {'shape': 'point',
+                                             'width': '0'}
+          for consumer in self._consumers[pcoll_id]:
+            producer_name = transform.unique_name
+            consumer_name = transforms[consumer].unique_name
+            self._edge_to_vertex_pairs[pcoll_id].append(
+                (producer_name, intermediate_dummy))
+            if self._pin:
+              edge_dict[(producer_name, intermediate_dummy)] = {
+                  'arrowhead': 'none',
+                  'label':
+                    self._pin.cacheable_var_by_pcoll_id(pcoll_id)}
+            else:
+              edge_dict[(producer_name, intermediate_dummy)] = {
+                  'arrowhead': 'none'}
+            self._edge_to_vertex_pairs[pcoll_id].append(
+                (intermediate_dummy, consumer_name))
+            edge_dict[(intermediate_dummy, consumer_name)] = {}
         else:
           for consumer in self._consumers[pcoll_id]:
             producer_name = transform.unique_name
             consumer_name = transforms[consumer].unique_name
             self._edge_to_vertex_pairs[pcoll_id].append(
                 (producer_name, consumer_name))
-            edge_dict[(producer_name, consumer_name)] = {}
+            if self._pin:
+              edge_dict[(producer_name, consumer_name)] = {
+                  'label':
+                    self._pin.cacheable_var_by_pcoll_id(pcoll_id)
+              }
+            else:
+              edge_dict[(producer_name, consumer_name)] = {}
```

Comment:
The entire logic will be simplified once data-centric user flow is in-place and pcollections are rendered as nodes rather than labels on edges.

The difference introduced:
Before:
![1](https://user-images.githubusercontent.com/4423149/68979012-364f6480-07b1-11ea-9ec3-8022154a8499.png)
After:
![2](https://user-images.githubusercontent.com/4423149/68979018-3cdddc00-07b1-11ea-8f62-6dd9a8a0e0ef.png)


In the future:
With data-centric user flow, notebook cell execution metadata and new Beam pipeline graph proposal:
![3](https://user-images.githubusercontent.com/4423149/68979139-93e3b100-07b1-11ea-8931-a6825b39b7a4.png)